k8s: Make Cilium aware of well known deployments such as kube-dns and etcd-operator

tgraf opened the pull request:
Make Cilium aware of kube-dns and the etcd-operator used for managed Kubernetes instead of requiring the user to assign fixed identities. This guarantees that existing label based kube-dns network policies continue working even in this new deployment mode. Previously, kube-dns would change its security relevant labels to `reserved:kube-dns`

Fixes: #5529

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5702)
<!-- Reviewable:end -->

nebril commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +59,86 @@ const (
+	ReservedEKSKubeDNS NumericIdentity = 102
 )
 
+type wellKnownIdentities struct {
```
Any specific reason reason for not going with

`type wellKnownIdentities map[NumericIdentity]wellKnownIdentity`
?

tgraf replied:
Yes, a struct allows being extended later on without changing all existing code.

ianvernon commented on Documentation/kubernetes/install/eks.rst:
```diff
@@ -54,21 +54,12 @@ Create EKS Cluster
        kubectl -n kube-system set env ds aws-node AWS_VPC_K8S_CNI_EXTERNALSNAT=true
 
-#. Assign a fixed security identity to ``kube-dns`` by  adding the label ``io.cilium.fixed-identity: kube-dns``
+#. Restart kube-dns/coredns to ensure that it is being managed by Cilium.
```
You should fix the README as well to be in sync with these new instructions - see commit https://github.com/cilium/cilium/commit/b1985eaf1a615a652a1d1c1978bc489873c6d9ab

tgraf replied:
EKS does not support coredns yet so it is not necessary. Referring to restart kube-dns is sufficient.

ianvernon commented on pkg/identity/cache_test.go:
```diff
@@ -34,6 +34,18 @@ func (s *IdentityTestSuite) TestLookupReservedIdentity(c *C) {
 	c.Assert(identity, Not(IsNil))
 	c.Assert(identity.ID, Equals, worldID)
+
+	initWellKnownIdentities()
```
Will we still retain the concept of "fixed-identity"? It seems like we will replace it with this PR?

tgraf replied:
Yes, they are part of the API and useful. We are just not using it for EKS anymore.

ianvernon commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +59,115 @@ const (
+	ReservedKubeDNS NumericIdentity = 102
+
+	// ReservedEKSKubeDNS is the reserved identity used for kube-dns on EKS
+	ReservedEKSKubeDNS NumericIdentity = 103
```
What is the need for a different identity for kube-dns on EKS? Can you elaborate as to why this is necessary?

EDIT: never mind, the labels are different below. I don't see a better way of doing this, but we are making things specific to implementations that are outside of our control (say Amazon decides to change kube-dns labels). It might be a good idea to mention the exactness of these labels being required to users in the EKS guide?

tgraf replied:
The user does not really have control over this but it definitely makes sense to document it 👍

ianvernon commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +59,115 @@ const (
+	}
+}
+
+// wellKnownIdentitity is an identity for well known security labels for which
```
well-known

ianvernon commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +59,115 @@ const (
+}
+
+// wellKnownIdentitity is an identity for well known security labels for which
+// a well known numeric identity is reserved to avoid requiring a cluster wide
```
well-known, cluster-wide

ianvernon commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +59,115 @@ const (
+
+// wellKnownIdentitity is an identity for well known security labels for which
+// a well known numeric identity is reserved to avoid requiring a cluster wide
+// setup. Examples of this includes kube-dns and the etcd-operator.
```
include

tgraf committed:
k8s: Make Cilium aware of well known deployments such as kube-dns and etcd-operator

Make Cilium aware of kube-dns and the etcd-operator used for managed Kubernetes
instead of requiring the user to assign fixed identities. This guarantees that
existing label based kube-dns network policies continue working even in this
new deployment mode. Previously, kube-dns would change its security relevant
labels to reserved:kube-dns

Fixes: #5529

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: André Martins <andre@cilium.io>
